src/schemapack/utils.py

```diff
@@ -0,0 +1,110 @@
+# Copyright 2021 - 2023 Universität Tübingen, DKFZ, EMBL, and Universität zu Köln
+# for the German Human Genome-Phenome Archive (GHGA)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Utility functions."""
+
+import json
+import os
+from collections.abc import Mapping
+from contextlib import contextmanager
+from pathlib import Path
+from typing import Any, Literal
+
+import jsonschema
+import jsonschema.exceptions
+import jsonschema.validators
+import yaml
+from jsonschema.protocols import Validator as JsonSchemaValidator
+
+
+class DecodeError(ValueError):
+    """Raised when decoding JSON or YAML data fails."""
+
+    def __init__(self, path: Path, assumed_format: Literal["JSON", "YAML"]):
+        super().__init__(
+            f"The file at '{path}' could not be decoded assuming {assumed_format}"
+            + " format."
+        )
+        self.path = path
+        self.assumed_format = assumed_format
+
+
+def read_json_or_yaml(path: Path) -> dict:
+    """Read JSON or YAML data from file. The format is decided based on the file
+    extension. If the file extension is not recognized, YAML is assumed (as it is a
+    superset of JSON).
+
+    Raises:
+        DecodeError: If the file cannot be decoded as JSON or YAML.
+    """
+    with path.open("rb") as file:
```

Comment:
I would use text mode here since these are text files, but it probably also works like this.

Reply:
Maybe add a test with some non-ascii chars to be sure.

Reply:
I also added an encoding. So now it should fail when non utf-8 characters are used.